test/unit/remotes_logs_test.rb

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+require 'test_helper'
+
+class ContainerLogsTest < KubernetesDeploy::TestCase
+  def test_print_latest_uses_prefix_if_multiple_containers
+    logs = KubernetesDeploy::RemoteLogs.new(parent_id: 'pod/pod-123-456', logger: logger,
+      container_names: %w(Container1 ContainerA))
+    logs.container_logs.first.expects(:lines).returns(mock_lines(1..4)).at_least_once
+    logs.container_logs.last.expects(:lines).returns(mock_lines('A'..'D')).at_least_once
+
+    logs.print_latest
+    assert_logs_match_all([
+      "[Container1]  Line 1",
+      "[Container1]  Line 2",
+      "[Container1]  Line 3",
+      "[Container1]  Line 4",
+      "[ContainerA]  Line A",
+      "[ContainerA]  Line B",
+      "[ContainerA]  Line C",
+      "[ContainerA]  Line D",
+    ], in_order: true)
+  end
+
+  def test_print_latest_does_not_use_prefix_if_one_container
+    logs = KubernetesDeploy::RemoteLogs.new(parent_id: 'pod/pod-123-456', container_names: %w(A), logger: logger)
+    logs.container_logs.first.expects(:lines).returns(mock_lines(1..4)).at_least_once
+
+    logs.print_latest
+    assert_logs_match_all(mock_lines(1..4), in_order: true)
+    refute_logs_match("[Container1]")
+  end
+
+  def test_print_all_prints_custom_message_if_no_logs_at_all
+    logs = KubernetesDeploy::RemoteLogs.new(parent_id: 'pod/pod-123-456', container_names: %w(A), logger: logger)
+    logs.print_all
+    assert_logs_match(%r{No logs found for pod/pod-123-456})
+  end
+
+  def test_print_all_prints_custom_message_if_one_container_has_no_logs
+    logs = KubernetesDeploy::RemoteLogs.new(parent_id: 'pod/pod-123-456', logger: logger,
+      container_names: %w(Container1 ContainerA))
+    logs.container_logs.first.expects(:lines).returns([]).at_least_once
+    logs.container_logs.last.expects(:lines).returns(mock_lines('A'..'D')).at_least_once
+
+    logs.print_all
+    assert_logs_match_all([
+      "No logs found for pod/pod-123-456 container 'Container1'",
+      "Logs from pod/pod-123-456 container 'ContainerA'",
+      "Line A",
+      "Line B",
+      "Line C",
+      "Line D"
+    ], in_order: true)
+  end
+
+  def test_print_all_identifies_logs_from_all_containers
+    logs = KubernetesDeploy::RemoteLogs.new(parent_id: 'pod/pod-123-456', logger: logger,
+      container_names: %w(Container1 ContainerA))
+    logs.container_logs.first.expects(:lines).returns(mock_lines(1..4)).at_least_once
+    logs.container_logs.last.expects(:lines).returns(mock_lines('A'..'D')).at_least_once
+
+    logs.print_all
+    assert_logs_match_all([
+      "Logs from pod/pod-123-456 container 'Container1'",
+      "Line 1",
+      "Line 2",
+      "Line 3",
+      "Line 4",
+      "Logs from pod/pod-123-456 container 'ContainerA'",
```

Comment:
Would this be better with a blank line printed between container logs?

Reply:
It does, but that's shitty to assert on because the full lines have timestamps in them. You can run these tests with `PRINT_LOGS=1` too btw:

![image](https://user-images.githubusercontent.com/4789493/46551883-ee1ffc00-c8a6-11e8-960c-ee680c05483c.png)

And hey, looks like I've actually got an _extra_ blank line in there!